designs/2023-only-run-reporting-rules/README.md

```diff
@@ -0,0 +1,127 @@
+- Repo: eslint/eslint
+- Start Date: 2023-01-16
+- RFC PR: <https://github.com/eslint/rfcs/pull/104>
+- Authors: Maddy Miller
+
+# Only run reporting lint rules
+
+## Summary
+
+Currently, ESLint will run all rules that are not marked as `off` in the configuration.
+This RFC proposes adding a way to configure which rules are actually run, to reduce linting
+time and better match the reporting outcome. Currently, when a rule is marked as `warn`,
+ESLint will still run the rule but not report the results when run under `--quiet`. The
+intended outcome of this RFC is to allow users to not run these rules when unnecessary.
+
+## Motivation
+
+A common use-case of the `warn` level is to report errors to the developer in their IDE,
+but to completely ignore it when running ESLint on the command line via the `--quiet` flag.
+When running with many rules set to the `warn` level, this can cause a significant amount
+of overhead on ESLint runtime. On an example large codebase with 154988 lines of TypeScript
+code, I found that by disabling all rules set to warn, I was able to reduce ESLint runtime
+from 4 minutes and 39 seconds to just 22 seconds. As the `warn` rules are entirely ignored
+in this situation, this is a substantial amount of wasted time and resources that could be
+avoided.
+
+## Detailed Design
+
+The proposed implementation is to modify the `--quiet` flag so that ESLint will skip
+running any rules that are set to the `warn` level.
+
+From an API perspective, this would be implemented by a predicate function that acts as
+a filter on which rules should be run. The predicate would take the rule configuration,
+and return whether the rule should be run. For simplicity, this function should not be
+given rules marked as `off`, as if this function handled existing behavior then users of the
+API would have to mimic that when attempting to extend it.
+
+The flag would be implemented by passing a predicate function that only returns true for
+rules that have been set to `error`, thus filtering out any that are marked as `warn`.
+
+The check for unused `eslint-disable` directives (`--report-unused-disable-directives`)
+should continue to mark `warn` rules as used, even when running with `--quiet`. As the
+rules are not actually run, an assumption would have to be made that all directives
+on rules marked `warn` are used when in this mode. This is a reasonable assumption, as
+the user likely does not expect `warn` flags to be touched at all in this mode.
+
+The `--max-warnings` flag should be disabled when the `--quiet` flag is in use, as it requires that
+`warn` rules are run to know the current number of reported warnings. This should provide an
+error when the two flags are used together.
```

Comment:
Per https://github.com/eslint/eslint/issues/14202, `--max-warnings` should be allowed with `--quiet`. Can we keep running rules if `--max-warnings` is specified?

Reply:
So that `--quiet` would have two different implementations depending on if `--max-warnings` is specified or not? I think this is a good idea since users who run those two flags together would have exact the same behavior after the feature is implemented and those who run `--quiet` without `--max-warnings` will have better performance.

Reply:
I've updated the RFC to cover this